### PR TITLE
Feature/refactor nitriding

### PR DESCRIPTION
**Pull request by mwittie**
This PR adds the ability to generate attestations based on the new interfaces. There is still a low of upstream code that needs to be refactored, so I'm simply plugging in our interfaces where necessary, but leaving other code unchanged.

### CODE REVIEW

**Review by ianhecker — CHANGES_REQUESTED** [review]

**Review by dlm — COMMENTED** [review]
I made it though the attestation package.  There are a lot of comments that correspond more broadly to the entire project.  Let's get those comments addressed (and fixed up across the project) and then we can take another pass

File: example/main_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+				assert.NoError(t, err)
+				assert.Contains(t, string(payload), tc.wantErr)
+				err = response.Body.Close()
+				assert.NoError(t, err)
+			}
+		})
+	}
+
+	err = attester.Close()
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply example/main_test.go:94]: done

File: attestation/verifier.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+type JSONVerifier struct {
+	publicKey *rsa.PublicKey
+}
+
+func MakeJSONVerifier(publicKey *rsa.PublicKey) JSONVerifier {
+	return JSONVerifier{publicKey: publicKey}
+}
+
+func (v JSONVerifier) Verify(signedDataBytes []byte) (SignedData, error) {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/verifier.go:21]: done

File: attestation/signer_test.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	"crypto/rsa"
+	"testing"
+
+	"github.com/blocky/nitriding/attestation"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestJSONSigner_HappyPath(t *testing.T) {
+	data := make([]byte, 20)
+	_, err := rand.Read(data)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/signer_test.go:14]: replaced with fuzzing

File: attestation/signer.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+type SignedData struct {
+	Data []byte `json:"Data"`
+	Sig  []byte `json:"Sig"`
+}
+
+type Signer interface {
+	Sign(data []byte) (signedDataBytes []byte, err error)
+	PublicKey() *rsa.PublicKey
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/signer.go:17]: updated

File: attestation/signer.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	"crypto/rsa"
+	"encoding/json"
+)
+
+type SignedData struct {
+	Data []byte `json:"Data"`
+	Sig  []byte `json:"Sig"`
+}
+
+type Signer interface {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/signer.go:15]: done

File: attestation/signer_test.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"testing"
+
+	"github.com/blocky/nitriding/attestation"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestJSONSigner_HappyPath(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/signer_test.go:12]: Renamed

File: attestation/checker_test.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func TestMakeStandaloneChecker_BadCertFile(t *testing.T) {
+	fileName := "empty_cert_file.pem"
+	file, err := os.Create(fileName)
+	assert.NoError(t, err)
+	err = file.Close()
+	assert.NoError(t, err)
+
+	_, err = attestation.MakeStandaloneChecker(fileName)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker_test.go:24]: done

File: attestation/checker_test.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	assert.Error(t, err)
+	assert.ErrorContains(t, err, "no such file")
+}
+
+func TestMakeStandaloneChecker_BadCertFile(t *testing.T) {
+	fileName := "empty_cert_file.pem"
+	file, err := os.Create(fileName)
+	assert.NoError(t, err)
+	err = file.Close()
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker_test.go:22]: addressed

File: attestation/checker_test.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMakeStandaloneChecker_NoCertFile(t *testing.T) {
+	_, err := attestation.MakeStandaloneChecker("no_such_file")
+	assert.Error(t, err)
+	assert.ErrorContains(t, err, "no such file")
+}
+
+func TestMakeStandaloneChecker_BadCertFile(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker_test.go:17]: solved

File: attestation/checker_test.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+	_, err = attestation.MakeStandaloneChecker(fileName)
+	assert.Error(t, err)
+	assert.ErrorContains(t, err, "could not append")
+
+	err = os.Remove(fileName)
+	assert.NoError(t, err)
+}
+
+func TestStandaloneChecker_CannotVerify(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker_test.go:32]: fixed

File: attestation/checker.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+func (c NitroChecker) CheckAttestationBytes(attestDoc []byte) (*nitrite.Result, error) {
+	return nitrite.Verify(attestDoc, nitrite.VerifyOptions{})
+}
+
+type StandaloneChecker struct {
+	roots *x509.CertPool
+}
+
+func MakeStandaloneChecker(rootCertFile string) (StandaloneChecker, error) {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker.go:36]: passing in a file, not a string now

File: attestation/checker.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/blocky/nitrite"
+)
+
+type Checker interface {
+	CheckAttestation(attest Attestation) (*nitrite.Result, error)
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker.go:14]: the two checkers behave pretty differently now, so I think the interface still makes sense

File: attestation/checker.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+	CheckAttestationBytes(attestBytes []byte) (*nitrite.Result, error)
+}
+
+type NitroChecker struct{}
+
+func (c NitroChecker) CheckAttestation(attest Attestation) (*nitrite.Result, error) {
+	attestDoc, err := base64.StdEncoding.DecodeString(string(attest))
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/checker.go:21]: solved

File: attestation/attester_test.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	"crypto/rsa"
+	"crypto/x509"
+	"os"
+	"testing"
+
+	"github.com/blocky/nitriding/attestation"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestStandaloneAttester_HappyPath(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/attester_test.go:14]: changed how the cert file gets passed around and cleaned up

File: attestation/attester.go [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+	err := a.session.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type StandaloneAttester struct {
+	signer     Signer
+	certFile   string
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/attester.go:78]: yes

File: attestation/attester.go [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+	"github.com/blocky/nitrite"
+	"github.com/fxamacker/cbor/v2"
+	"github.com/hf/nsm"
+	"github.com/hf/nsm/request"
+	"github.com/veraison/go-cose"
+)
+
+type Attestation string
+
+type Attester interface {
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/attester.go:24]: It does have a session, which is why I added the close function. Open to suggestions

File: attestation/attester.go [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+	return NitroAttester{
+		session: session,
+		signer:  signer,
+	}, nil
+}
+
+func (a NitroAttester) Attest(nonce, userData []byte) (Attestation, error) {
+	signedUserData, err := a.signer.Sign(userData)
+	if err != nil {
+		return "", err
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/attester.go:49]: Its a string, but I made `Attestation` conform to the stringer interface, so we don't have to cast it anymore

File: attestation/attester.go [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+type Attester interface {
+	Attest(nonce, userData []byte) (attestation Attestation, err error)
+	Close() error
+}
+
+type NitroAttester struct {
+	session    *nsm.Session
+	signer     Signer
+	privateKey *rsa.PrivateKey
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation/attester.go:32]: addressed

File: attestation.go [hunk 19]
```diff
@@ -2,14 +2,10 @@ package nitriding
...
 	"bytes"
-	"encoding/base64"
-	"encoding/hex"
 	"errors"
 	"fmt"
-	"net/http"
-	"regexp"
 
-	"github.com/hf/nitrite"
+	"github.com/blocky/nitrite"
```
[the remark this answers is not in the document]
Comment (mwittie) [reply attestation.go:8]: nitrite is not a part of this repo and following our earlier discussion I didn't want to embed a nitrite submodule. So, without a submodule, I think this is the alternative of getting the right version of the code into the project.

**Pull request closed without merging**